src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -812,23 +814,31 @@ private UsingDirectiveSyntax ParseUsingDirective()
                 //NB: there's no way this could be true for a set of tokens that form a valid 
                 //using directive, so there's no danger in checking the error case first.
 
-                name = WithAdditionalDiagnostics(CreateMissingIdentifierName(), GetExpectedTokenError(SyntaxKind.IdentifierToken, this.CurrentToken.Kind));
+                type = WithAdditionalDiagnostics(CreateMissingIdentifierName(), GetExpectedTokenError(SyntaxKind.IdentifierToken, this.CurrentToken.Kind));
                 semicolon = SyntaxFactory.MissingToken(SyntaxKind.SemicolonToken);
             }
             else
             {
-                name = this.ParseQualifiedName();
-                if (name.IsMissing && this.PeekToken(1).Kind == SyntaxKind.SemicolonToken)
-                {
-                    //if we can see a semicolon ahead, then the current token was
-                    //probably supposed to be an identifier
-                    name = AddTrailingSkippedSyntax(name, this.EatToken());
-                }
+                // In the case where we don't have an alias, only parse out a name for this using-directive.  This is
+                // worse for error recovery, but it means all code that consumes a using-directive can keep on assuming
+                // it has a name when there is no alias.  Only code that specifically has to process aliases then has to
+                // deal with getting arbitrary types back.
+                type = alias == null ? this.ParseQualifiedName() : this.ParseType();
+
+                if (type is RefTypeSyntax)
+                    type = AddError(type, ErrorCode.ERR_BadRefInUsingAlias);
+
+                // If we can see a semicolon ahead, then the current token was probably supposed to be an identifier
+                if (type.IsMissing && this.PeekToken(1).Kind == SyntaxKind.SemicolonToken)
+                    type = AddTrailingSkippedSyntax(type, this.EatToken());
+
+                if (unsafeToken != null && alias == null)
+                    unsafeToken = AddError(unsafeToken, ErrorCode.ERR_BadUnsafeInUsingDirective);
```

Comment:
nit: I assumed you were inclined to move parsing errors (other than missing token and unexpected token) out to binding. Just curious, why did you prefer parser errors for these new errors?

Reply:
Good point.  I'll move :)

Reply:
Moved.